introc/c_memory_model.tex

```diff
@@ -56,37 +55,50 @@ \subsection{Structs}
-/*
-Currently, our memory looks like this with junk in those black spaces
-                ___ ___ ___ ___ ___ ___ ___ ___ ___ ___ ___
- bhuvan_name = |___|___|___|___|___|___|___|___|___|___|___|
+\end{lstlisting}
+
+Currently, our memory looks like the following image. There is nothing in those boxes
 
-*/
+\begin{figure}[H]
+\centering
+\includegraphics[width=.7\textwidth]{introc/drawings/memory_model_empty.png}
+\caption{Struct pointing to 11 empty boxes}
+\end{figure}
 
+So what happens when we assign length?
+The first four boxes are filled with the value of the variable at length.
+The rest of the space is left untouched.
 
+\begin{lstlisting}[language=C]
 bhuvan_name->length = length;
-/*
-This writes the following values to the first four bytes
-The rest is still garbage
-                ___ ___ ___ ___ ___ ___ ___ ___ ___ ___ ___
- bhuvan_name = | 0 | 0 | 0 | 6 |___|___|___|___|___|___|___|
+\end{lstlisting}
 
-*/
+\begin{figure}[H]
+\centering
+\includegraphics[width=.7\textwidth]{introc/drawings/memory_model_length.png}
+\caption{Struct pointing to 11 boxes, 4 filled with 0006, 7 junk}
+\end{figure}
 
+Now, we can write a string to the end of our struct with the following call.
 
+\begin{minted}{C}
 strcpy(bhuvan_name->c_str, to_convert);
-/*
-Now our string is filled in correctly at the end of the struct
+\end{minted}
 
-                ___ ___ ___ ___ ___ ___ ___ ___ ___ ___ ____
- bhuvan_name = | 0 | 0 | 0 | 6 | b | h | u | v | a | n | \0 |
-                                                           ‾
-*/
+\begin{figure}[H]
+\centering
+\includegraphics[width=.7\textwidth]{introc/drawings/memory_model_full.png}
+\caption{Struct pointing to 11 boxes, 4 filled with 0006, 7 the stirng ``bhuvan''}
+\end{figure}
 
+We can even do a sanity check to make sure that the strings are equal.
+
+\begin{minted}{C}
 strcmp(bhuvan_name->c_str, "bhuvan") == 0 //The strings are equal!
-\end{lstlisting}
+\end{minted}
 
-What that zero length array does is point to the \textbf{end of the struct} this means that the compiler will leave room for all of the elements calculated with respect to their size on the operating system (ints, chars, etc). The zero length array will take up no bytes of space.
+What that zero length array does is point to the \textbf{end of the struct} this means that the compiler will leave room for all of the elements calculated with respect to their size on the operating system (ints, chars, etc).
+The zero length array will take up no bytes of space.
 Since structs are just continuous pieces of memory, we can allocate \textbf{more} space than required and use the extra space as a place to store extra bytes.
 Although this seems like a parlor trick, it is an important optimization because to have a variable length string any other way, one would need to have two different memory allocation calls.
 This is highly inefficient for doing something as common in programming as is string manipulation.
```

Comment:
Line 109: "when you have to an address from memory"  => "when you have to load an address from memory"